builtin/logical/database/dbplugin/plugin.go

```diff
@@ -0,0 +1,138 @@
+package dbplugin
+
+import (
+	"fmt"
+	"net/rpc"
+	"time"
+
+	"github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/vault/helper/pluginutil"
+	log "github.com/mgutz/logxi/v1"
+)
+
+// DatabaseType is the interface that all database objects must implement.
+type DatabaseType interface {
+	Type() (string, error)
+	CreateUser(statements Statements, usernamePrefix string, expiration time.Time) (username string, password string, err error)
+	RenewUser(statements Statements, username string, expiration time.Time) error
+	RevokeUser(statements Statements, username string) error
+
+	Initialize(config map[string]interface{}, verifyConnection bool) error
+	Close() error
+}
+
+// Statements set in role creation and passed into the database type's functions.
+type Statements struct {
```

Comment:
Based on issues we've had in the past with delimited strings being passed in, I think we should do something different. For context:

```
Statements to be executed to create a user. Must be a semicolon-separated
 +							string, a base64-encoded semicolon-separated string, a serialized JSON string
 +							array, or a base64-encoded serialized JSON string array. The '{{name}}', 
 +							'{{password}}', and '{{expiration}}' values will be substituted.
```
We did this base64 stuff because we didn't have a better way at the time, but now we have a TypeStringSlice and TypeCommaStringSlice (and could easily add TypeSemicolonStringSlice). It feels like a less jank way to do this is to simply allow the user to submit either a single string semicolon-delimited or a slice of strings, then store the string array instead.

@vishalnayak IIRC base64 was so that there could be semicolons within the statements that weren't broken; I don't remember however why we had a base64-encoded JSON string array other than, I guess, to allow people to avoid escaping quotes in their JSON. Do you remember if there was any other reason? Not saying that we couldn't keep base64 support, I just don't know that it is really necessary since most people using JSON won't hand-encode it but will likely rather feed a string into a serializer which will escape as necessary. The only other thing I can think of is for mixing " and ' within statements...but again, escaping can be used, or we could still allow base64 support via e.g. a TypeDatabaseStatements.